prusti-specs/src/specifications/preparser.rs

```diff
@@ -289,557 +58,547 @@ impl Parse for SpecEntArgs {
-            // or nothing at all after the parenthesized block, then we might parser
-            // this as a Prusti assertion
+    fn parse_rust_until(&mut self, terminator: &str) -> syn::Result<ExpressionWithoutId> {
+        let mut t = vec![];
 
-            if self.expected_operator {
-                return Err(self.error_expected_operator());
-            }
+        while !self.peek_operator("|=") &&
+              !self.peek_operator("&&") &&
+              !self.peek_operator("==>") &&
+              !self.peek_operator(terminator) &&
+              !self.tokens.is_empty() {
+            t.push(self.pop().unwrap());
+        }
+        let mut stream = TokenStream::new();
+        stream.extend(t.into_iter());
 
-            let mut parser = Parser::from_token_stream(group.stream());
-            let conjunct = parser.extract_assertion();
+        let cloned: VecDeque<TokenTree> = stream.clone().into_iter().collect();
+        if let Some(span) = self.contains_operator_recursive(&cloned, "==>") {
+            Err(self.error_no_implies(span))
+        } else if cloned.is_empty() {
+            Err(self.error_expected("expression"))
+        } else {
+            Ok(ExpressionWithoutId {
+                spec_id: common::SpecificationId::dummy(),
+                id: (),
+                expr: syn::parse2(stream)?,
+            })
+        }
+    }
 
-            if let Err(err) = conjunct {
-                return Err(err);
+    /// is there any non-prusti operator following the first thing?
+    fn is_part_of_rust_expr(&mut self) -> bool {
+        if let Some(token) = self.tokens.pop_front() {
+            if self.peek_operator("|=") ||
+               self.peek_operator("&&") ||
+               self.peek_operator("==>") ||
+               self.tokens.front().is_none() {
+                self.tokens.push_front(token);
+                false
+            } else {
+                self.tokens.push_front(token);
+                true
             }
-
-            let conjunct = conjunct.unwrap();
-            self.conjuncts.push(conjunct);
-            self.previous_expression_resolved = true;
-            self.expected_operator = true;
-        }
-        else{
-            // if this was not the case, we just parse as a Rust expression and stick
-            // it to any possible already being-parsed expression
-            self.expr.push(TokenTree::Group(group));
+        } else {
+            false
         }
-        Ok(())
     }
-    fn resolve_rust_expr(&mut self) -> syn::Result<()> {
-        // if nothing of the above happened, we just continue parsing as a Rust expression
-        if self.expected_only_operator {
-            self.input.pop();
-            return Err(self.error_expected_operator());
+    /// does the given operator appear in the stream at top level
+    fn contains_operator(&self, stream: &VecDeque<TokenTree>, operator: &str) -> bool {
+        let mut stream = stream.clone();
```

Comment:
It's not very nice that this read-only operation has to clone the stream.

Reply:
I'll look into this; I think I would then have to make peek_operator_stream take an offset, and then iterate through by index.

Reply:
Addressed by c77b6d0b1638c36b81eddb1e87af37c3f1fff212